schema/query/projection_evaluator.go

```diff
@@ -79,47 +214,57 @@ func evalProjection(ctx context.Context, p Projection, payload map[string]interf
 					}
 				} else if ref, ok := def.Validator.(*schema.Reference); ok {
```

Comment:
nicer code with a type-switch btw.

```go
switch ref := def.Validator.(type) {
case *schema.Reference:
     // ref will be of type  *schema.Reference in this block
case *schema.Array:
    //  ref will be of type  *schema.Array in this block
case FieldGetter:
    //  ref will be of type  FieldGetter in this block.
default:
   return nil, fmt.Errorf("%s: field has no children", pf.Name)
}
```

Slightly better to have `FieldGetter` than `*schema.Dict`; this way we can also handle `*schema.Object`. `*schema.Array` is a bit different, unfortnunatly, and should probalby implement som other interface, e.g.

```go
type IndexGetter interface{
    GetField(i int) (Field, bool)
}
```

But we can leave that for later. A static `*schema.Array` check is fine for now.

I must also note though that this function in general (as it was written before this review), is too long and to deeply nested to my taste, and could benefit from some helper functions, or re-evaluation of the if/else logic/nesting. That's for later though.

Reply:
Made suggested type switch in `evalProjectionArray`.